### ISSUE TITLE
Allow users to pass an optional dictionary of placeholders, values that is added to feed_dict in utils_tf methods (e.g., training, eval, ...)

### ISSUE THREAD

**Issue by kirk86** [tensorflow/cleverhans#142]
Hi I was wondering if it is possible to remove dependency of `K.set_learning_phase(0)` in the tutorials?
 If I try to remove it I get all sorts of errors. If it is removed none of the attacks works. Mainly in the tutorials if we replace `model_train` with `model.fit` and `model_eval` with `model.evaluate` everything works fine without the need of `K.set_learning_phase()` except from the actual attacks. Any idea why is this happening?

**Comment by npapernot** [comment]
Hi, the keras learning phase variable is used to control layers that have different behavior at training and test time (e.g., dropout). In the tutorials, we don't use these layers. To avoid having to feed a value of this variable in all TF graph runs performed by cleverhans, we simply set it to 0. If you use K.fit, you may not be able to use some of the cleverhans functions such as adversarial training.

**Comment by kirk86** [comment]
@npapernot thanks a lot for the prompt respond. Yeah actually I've had dropout and batch norm layers and couldn't figure out what was going on.

> If you use K.fit, you may not be able to use some of the cleverhans functions such as adversarial training

IMHO I think it might be doable, unless I'm missing sth else fundamental here. So if I do K.fit for training I can say sth like this along the lines when I compile the model before training:

```
# Define loss
loss = model_loss(y, predictions)
loss = (loss + model_loss(y, predictions_adv)) / 2
K.compile(optimizer='adam', loss=loss, metrics=['accuracy']) 
```

**Comment by npapernot** [comment]
Sorry for the confusion, yes I meant it will be hard to use cleverhans functions "as is" because the learning phase will raise a value not fed error. 

Actually, I had a related thought the other day: we may want to have an optional argument to cleverhans functions that run a TF graph where users can specify a feeding dictionary of additional variables along with their values that will be appended to the feeding dictionary hard-coded in cleverhans (when running `sess.run`.

**Comment by kirk86** [comment]
@npapernot 

> Sorry for the confusion, yes I meant it will be hard to use cleverhans functions "as is" because the learning phase will raise a value not fed error.

Yes indeed, I had this error the other day and was trying to figure out what was going. If someone removes `K.set_learning_phase()` will definitely need to substitute that in the session feed dictionary as {'K.set_learning_phase()': 0} or equivalently {'keras_learning_phase': 0}.

> Actually, I had a related thought the other day: we may want to have an optional argument to cleverhans functions that run a TF graph where users can specify a feeding dictionary of additional variables along with their values that will be appended to the feeding dictionary hard-coded in cleverhans (when running sess.run)

This is much better and makes the functionality of the library more flexible this way. For instance one might just need to use some functions and not change the whole code base from K.fit to model_train when appropriate.

**Comment by goodfeli** [comment]
I'm hoping to write a tutorial that uses only TensorFlow, with no framework
on top. That would remove all dependence on Keras. But it might take a
while before I get to this, so someone else is welcome / encouraged to do
it too.

**Comment by kirk86** [comment]
Hey guys, just a quick question/extension on what @npapernot suggested about `feed_dict`. I've noticed that all functions take params such `train_params` or `eval_params` so I tried to remove `K.set_learning_phase(0)` and hard coded it in the params as `'keras_learnig_phase': 0` but after doing some digging up I found out that `batch_eval` is not using/passing all arguments to feed_dict, so there is no current way to pass that argument to the tensorflow feed_dict. Maybe this should be sth to be considered regarding what @npapernot suggested about allowing users to pass arguments. So at the moment we can pass arguments to the different functions but the `_FlagsWrapper object` that does the argument processing return only a number of those which are later fed to `feed_dict`. Was that intentional or by desing?

**Comment by npapernot** [comment]
Yes, that was what I wanted to provide as a new feature: to allow people to provide custom placeholders and their values that should be added to the feeding dictionary when running the session.

The current argument processing was not designed to support this, but it was not intentional.
